Keep visual mode for some motions

daniloisr opened the pull request:
Keep visual mode for the following motions:
- ctrl-f
- ctrl-b
- ctrl-d
- ctrl-u

Solves #593 and #614

daniloisr commented:
There is some other way to get the editor height without this code?

```coffee
firstScreenRow = @editor.getFirstVisibleScreenRow()
lastScreenRow = @editor.getLastVisibleScreenRow()
height = lastScreenRow - firstScreenRow
```

daniloisr commented:
I think that it's ok now :v:

jacekkopecky commented:
@daniloisr this PR changes the behavior of these commands - in VIM they scroll the screen so the cursor stays on the same line on the screen, here they don't. I think we'd want to keep the scrolling behavior as well as fixing how these motions don't affect the visual mode. (Yes, they do seem to be motions rather than scrolls.)

daniloisr commented:
@jacekkopecky you are right friend, sorry about that, I have used ctrl-d and ctrl-u for a long time thinking that they are motions, but I read the docs right now and saw that they are scrolls.

I could try to fix they as scrolls, but for now, what you think about I letting this PR only for ctrl-f and ctrl-b fix?

daniloisr commented:
Ops, ctrl-f and ctrl-b are scroll too, :/

I already started working on that, soon I will fix this PR

daniloisr commented:
@jacekkopecky I have changed the code and added visual selection to `Scroll` class*, and before I move to `ctrl-f` and `ctrl-b` implementation I would like to know your opinion, did you think that this way is correct?

\* I used the `select` method from `Motion` class, so maybe its better to join these codes... I'm studying it now

jacekkopecky commented:
Hi @daniloisr, this looks interesting, I'm hoping to get enough time to focus on this in a few days, hope that's OK.

daniloisr commented:
Finished this PR.

I think that after this PR I will work on other related problems like keeping selection when scrolling the whole screen with `ctrl-y`, because today we lose the selection.

maxbrunsfeld commented:
@daniloisr Thanks for working on this! Because these commands move the cursor and modify the selection, I would really prefer to implement them as subclasses of `Motion`. That way we wouldn't have to duplicate the logic for updating the selection.

You should be able to implement `Motion::moveCursor(cursor)` so that it positions the cursor correctly using `Cursor::setBufferPosition` or `::setScreenPosition`, and then override `Motion::select` and `Motion::execute` so that they update the `@editor`'s scroll position at the end.

Let me know if that approach makes sense to you, I can give you more detailed guidance if it'd be helpful.

maxbrunsfeld commented on spec/scroll-spec.coffee:
```diff
@@ -121,31 +121,79 @@ describe "Scrolling", ->
+        keydown('d', ctrl: true)
+        expect(editor.getSelectedText()).toEqual [42..52].join("\n").concat("\n")
+
+    fdescribe "the ctrl-f keybinding", ->
```
Same here.

maxbrunsfeld commented on spec/scroll-spec.coffee:
```diff
@@ -121,31 +121,79 @@ describe "Scrolling", ->
+        keydown('u', ctrl: true)
+        expect(editor.getSelectedText()).toEqual [32..42].join("\n").concat("\n")
+
+    fdescribe "the ctrl-b keybinding", ->
```
Gotta remove the `f` here before we merge, too.

daniloisr commented:
Hi @maxbrunsfeld. My first approach was to create a `Motion` subclass, but these commands also scrolls the screen (like @jacekkopecky said) and we have the [`Scroll` class that defines some useful stuffs like `@scrolloff` and `@rows`](https://github.com/atom/vim-mode/blob/master/lib/scroll.coffee#L5), so if I create a `Motion` subclass I will need them too.

I was thinking a lot about it, and my solution was extract the selection logic to a mixin because I think that `Motion` classes can use this mixin later (I can implement it next).

Do you think that implementing a `Motion` subclasse still a better solution? :sweat_smile:

maxbrunsfeld commented:
Do you still have your commits where you wrote it as a `Motion` subclass? I don't see them on this branch. You could just push them to a different branch if so; I'm curious what it looked like.

> we have the Scroll class that defines some useful stuffs like @scrolloff and @rows
That's true. I just think the `Motion` class defines a lot *more* useful stuff than the `Scroll` class, and I think `Motion` is a more correct way to think of these commands, since they can be composed with operators.

daniloisr commented:
@maxbrunsfeld sorry, I deleted the branch =P. But dont worry, the solution was almost like [the class `MoveToBottomOfScreen`](MoveToBottomOfScreen), I just did the right math, when @jacekkopecky commented I had not implemented the scrolling yet.

I liked your idea, soon I can write a new code doing the scroll in a `Motion` subclass. What you think?

maxbrunsfeld commented:
:+1: Sounds good. Sorry for not giving you feedback earlier. I definitely want to merge this fix.

daniloisr commented:
@maxbrunsfeld done! Later I will try to improve the way the selection is made, I have some felling that it may be simpler.

The build has failed due module download problem, I tried to rerun the build but I dont know how.

maxbrunsfeld commented:
Looks *really* nice. Re-running the build now.

maxbrunsfeld commented on lib/motions/general-motions.coffee:
```diff
@@ -379,11 +379,56 @@ class MoveToMiddleOfScreen extends MoveToScreenLine
+
+  select: (count, options) ->
+    @scrollScreen(count)
+    super(count, options)
```
So I just tried out your changes. I noticed that when in visual mode, the cursor's position is not preserved relative to the screen. I think you may need to call `@scrollScreen()` *after* calling `super` here:

```coffee
result = super
@scrollScreen(count)
result
```

daniloisr replied:
@maxbrunsfeld fixed it storing the `destination` of  `setScrollTop` before calling `super` and setting the `@editor.setScrollTop` again. I know that this is not the best solution but this happen because the `Motion.select` is changing the scroll value, I tried to sove this changing the `moveSelectionLinewise` and `moveSelectionInclusively` code but I could find a better solution.

maxbrunsfeld commented on lib/motions/general-motions.coffee:
```diff
@@ -379,11 +379,56 @@ class MoveToMiddleOfScreen extends MoveToScreenLine
+
+  execute: (count) ->
+    @scrollScreen(count)
+    super(count)
```
You may need to do the same thing here, but I'm not sure.

daniloisr commented:
@maxbrunsfeld I think it's ok now o/

maxbrunsfeld commented on lib/motions/general-motions.coffee:
```diff
@@ -379,11 +379,62 @@ class MoveToMiddleOfScreen extends MoveToScreenLine
+  currentFirstScreenRow: 0
+
+  constructor: (@editor, @vimState) ->
+    super(@editor, @vimState)
```
You can remove this `constructor` definition, since it only calls `super`.

maxbrunsfeld commented on lib/motions/general-motions.coffee:
```diff
@@ -379,11 +379,62 @@ class MoveToMiddleOfScreen extends MoveToScreenLine
+    @editor.setScrollTop(final_destination)
+
+  execute: (count) ->
+    final_destination = @scrollScreen(count)
```
We usually don't use `snake_case` variable names. I would say call this variable `scrollDestination`.

maxbrunsfeld commented:
Looks great! Other than some minor comments, I'll merge.

daniloisr commented:
@maxbrunsfeld thats it, thx for the review =)
